### PR TITLE
use buffer interface in write_direct_chunk

### PR DESCRIPTION
**Pull request by weninc**
#1519

### CODE REVIEW

File: h5py/h5d.pyx [hunk 1]
```diff
@@ -405,7 +409,7 @@ cdef class DatasetID(ObjectID):
 
     IF HDF5_VERSION >= (1, 8, 11):
 
-        def write_direct_chunk(self, offsets, bytes data, filter_mask=0x00000000, PropID dxpl=None):
+        def write_direct_chunk(self, offsets, data, filter_mask=0x00000000, PropID dxpl=None):
```
Comment (scopatz) [inline h5py/h5d.pyx]: we probably shouldn't change the API  like this

Comment (weninc) [reply h5py/h5d.pyx]: But this only relaxes the type constraint and it doesn't change how the function is called. So it should be transparent to anyone using the API and calling it with bytes still works without any change. It just allows other types as well

**Comment by takluyver** [comment]
I actually think that allowing numpy arrays directly here is a *bad* idea - if you can pass an ndarray, some people will assume that it takes account of the metadata, e.g. doing automatic conversions based on the dtype, or using the array shape somehow, and then get confused when the 'wrong' data is stored.

However, you can get a memoryview of an array's data with `arr.data`, without copying it. So maybe a suitable middle-ground is to allow bytes or memoryviews?

**Comment by weninc** [comment]
Its a very low level function, so I don't think people would expect the numpy metadata would be accounted for. You have to be very careful with this function anyway especially with compressed data. We are trying to use this to store large amounts of precompressed data from large image detectors and we can't afford copying our data into bytes objects.
Memoryviews of arr.data would be fine for us, but I tried them and can't get it to work:
```
dset.id.write_direct_chunk((i, 0, 0), data.data)
TypeError: Argument 'data' has incorrect type (expected bytes, got memoryview)
```

**Comment by takluyver** [comment]
Fair enough, I guess the buffer interface is the obvious way to do this. But I'd like it to be noted in the docstring that passing an array just saves the raw data from the array, ignoring the shape and dtype completely. The docstring is also used in the low-level API docs at https://api.h5py.org/ .

The CI is also asking you to add yourself to the `.authors.yml` file.

File: h5py/h5d.pyx (hunk 1 above)
Comment (scopatz) [reply h5py/h5d.pyx]: But it removes a positional argument, which causes all downstream calls of this function to break :cry:

Comment (takluyver) [reply h5py/h5d.pyx]: @scopatz I think you're misreading the change - it's not removing an argument, just removing the `bytes` type from the `data` argument (a bit of C like Cython syntax).

**Comment by weninc** [comment]
Great thanks. I updated the docstring to reflect the ndarry metadata issue and added myself to the .authors.yml file. Let me know if you need anything else.

File: h5py/h5d.pyx (hunk 1 above)
Comment (scopatz) [reply h5py/h5d.pyx]: Oh! I am! Sorry.  Serves me right for trying to reveiw on my phone

File: h5py/h5d.pyx [hunk 2]
```diff
@@ -405,13 +409,18 @@ cdef class DatasetID(ObjectID):
...
+            """ (offsets, data, uint32_t filter_mask=0x00000000, PropID dxpl=None)
 
             This function bypasses any filters HDF5 would normally apply to
             written data. However, calling code may apply filters (e.g. gzip
             compression) itself before writing the data.
 
+            `data` is a Python object that implements the Py_buffer interface.
+            In case of a ndarray the shape and dtype are ignored. It's the
+            users responsibility to make sure they are compatible with the
```
Comment (takluyver) [inline h5py/h5d.pyx]:
```suggestion
            user's responsibility to make sure they are compatible with the
```

Sorry to nitpick :wink:

**Comment by takluyver** [comment]
Thanks! Other than that tiny change, LGTM. I'll give it a day or two in case anyone else wants to review it, then I'll merge it if not.

File: h5py/h5d.pyx (hunk 2 above)
Comment (weninc) [reply h5py/h5d.pyx]: oh yes of course :sweat_smile:

**Comment by kif** [comment]
Hi Clemens,

I appreciate very much your contribution as this is something we are also struggling with.

The H5DOwrite_chunk is called in a *nogil* environment, hence the  Py_buffer.len and Py_buffer.buf should be Python free. According to the doc, it is the case.

+1 for merging this PR.

**Pull request merged**